core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1297,11 +1332,188 @@ public TimeValue getInactiveTime() {
+
+        final String name;
+        volatile TimeValue interval;
+        volatile ScheduledFuture<?> scheduledFuture;
+
+        public ScheduledTask(String name, TimeValue interval) {
+            this.name = name;
+            this.interval = interval;
+            this.scheduledFuture = null;
+        }
+
+        public void startIfNeeded() {
+            synchronized (mutex) {
+                if (scheduledFuture != null) {
+                    return;
+                }
+                if (shouldRun()) {
+                    logger.debug("scheduling [{}] every [{}]", name, interval);
+                    reschedule();
+                } else {
+                    logger.debug("scheduled [{}] disabled", name);
+                }
+            }
+        }
+
+        /** returns true if this task should be running for this shard and setting */
+        protected boolean shouldRun() {
+            synchronized (mutex) {
+                return interval.millis() > 0 && shouldRunStates().contains(state) && shardRouting.primary();
+            }
+        }
+
+        public void cancelAndStartIfNeeded() {
+            synchronized (mutex) {
+                cancel();
+                startIfNeeded();
+            }
+        }
+
+        public void cancel() {
+            synchronized (mutex) {
+                FutureUtils.cancel(scheduledFuture);
+                scheduledFuture = null;
+            }
+        }
+
+        @Override
+        public void onFailure(Throwable t) {
+            if (t instanceof AlreadyClosedException || t instanceof EngineClosedException || state == IndexShardState.CLOSED) {
+                logger.trace("ignoring unexpected error while running [{}], shutting down", t, name);
+            } else {
+                logger.warn("unexpected error while running [{}]", t, name);
+                reschedule();
+            }
+        }
+
+        final EnumSet<IndexShardState> DEFAULT_RUNNING_STATES = EnumSet.of(IndexShardState.RECOVERING,
+                IndexShardState.POST_RECOVERY, IndexShardState.STARTED, IndexShardState.RELOCATED);
+
+        protected EnumSet<IndexShardState> shouldRunStates() {
+            return DEFAULT_RUNNING_STATES;
+        }
+
+        /**
+         * Schedules another (future) task, if refresh_interval is still enabled.
+         */
+        protected void reschedule() {
+            synchronized (mutex) {
+                if (shouldRun()) {
+                    scheduledFuture = threadPool.schedule(interval, ThreadPool.Names.SAME, this);
+                }
+            }
+        }
+    }
+
+    class GlobalCheckpointUpdater extends ScheduledTask {
+
+        public GlobalCheckpointUpdater(TimeValue interval) {
+            super("checkpoint update", interval);
+        }
+
+        @Override
+        protected boolean shouldRun() {
+            synchronized (mutex) {
+                return super.shouldRun() && shardRouting.primary();
```

Comment:
I don't understand this condition. The base implementation already includes ``&& shardRouting.primary()``?